### ISSUE TITLE
Allow OR for requirements (or reintroduce nreqs)

### ISSUE THREAD

**Issue by Corbeau-1** [longturn/freeciv21#2326]
Currently, there is only AND operator possible for effects (and buildings and techs and...) When vectors are listed, the effect is... in effect... only when all requirement are fulfilled. So, there is AND, but there is no OR. The result is that some effects need to be split into more [effects], or a whole block of exclusions needs to be listed. A classic example is +1 trade for Republic and Democracy: right now there need to be two effects, one for Rep and Demo each, or you have to list all governments and exclude all non-Republic and non-Democracy. With the OR operator, there could be one simple effect: "have +1 trade for Republic OR Democracy".

Additionally, some features are not possible to implement. A completely random example: some buildings or units could be made possible with the discovery of any of the listed techs. This would allow for more diverse research paths, more diverse gameplay etc. etc. I could probably think of dozens of other examples.

Technically, I don't know how I'd even start formatting this, but the OR operator can be allowed more simply by reintroducing `nreqs` which were removed, I think, somewhere between Freeciv 2.4 and 2.6.

**Comment by XHawk87** [comment]
Possibly the simplest way to implement this on top of the current system might be to allow defining multiple `reqs` fields. If all conditions of any of the `reqs` fields are met, the effect/unit/action/building/tech is allowed. 

E.g. allow defining units that can be captured only if their hitpoints drop below a threshold
```
[actionenabler_capture_units]
action = "Capture Units"
actor_reqs    =
    { "type",         "name",     "range", "present"
      "UnitFlag",     "Capturer", "Local", TRUE
      "DiplRel",      "War",      "Local", TRUE
      "MinMoveFrags", "1",        "Local", TRUE
      "DiplRel",      "Foreign",  "Local", TRUE
    }
target_reqs    =
    { "type",         "name",           "range", "present"
      "UnitFlag",     "Capturable", "Local", TRUE
    }
target_reqs    =
    { "type",         "name",           "range", "present"
      "UnitFlag",     "WeakCapturable", "Local", TRUE
      "MinHitPoints", "5",              "Local", FALSE
    }
```
Currently, the only way to add this is to have it as a separate action enabler. This means that in situations where you have a mixture of Capturable and WeakCapturable units on low hitpoints on the same tile, both action enablers will fail on detecting the other kind of capturable units and disallow the action. The desired outcome is that the whole lot are captured together and an effective OR clause is the only way to define that.

**Comment by lmoureaux** [comment]
Unfortunately the second `target_reqs` would overwrite the first at a very low level if they have the same name.

**Comment by XHawk87** [comment]
Well, if it's at a low level then changing it could potentially break all kinds of things. Not ideal, but would `target_reqs, target_reqs2, ..., target_reqsn` be better?

**Comment by lmoureaux** [comment]
We probably don't want to change the secfile reader to deal with this. `target_reqs2` (or `target2_reqs`?) would work.

**Comment by Corbeau-1** [comment]
>  `target_reqs2` (or `target2_reqs`?) would work.

Would work now? Without changing the code?

**Comment by XHawk87** [comment]
> > `target_reqs2` (or `target2_reqs`?) would work.
> 
> Would work now? Without changing the code?

No, it still requires code changes, but the changes would be less major and less likely to break things.

**Comment by XHawk87** [comment]
Ideally, I'd like us to eventually be able to move away from secfiles to something more standardised. Defining rules in Lua #2514, and using JSON for data serialisation and interchange. Then it could look something like this:
```lua
function unit_is_capturable(unit)
  if unit.utype:has_flag("Capturable") then return true end
  if unit.utype:has_flag("WeakCapturable") then
    return unit.hitpoints < 5
  end
  return false
end

action.create(function(a)
  a.allow_when(function(unit, tile)
    if not unit.utype:has_flag("Capturer") then return false end
    if unit.move_fragments < 1 then return false end
    for target_unit in tile:units_iterate() do
      if target_unit.owner == unit.owner then return false end
      if unit.owner:diplstate(target_unit.owner) ~= "War" then return false end
      if not unit_is_capturable(target_unit) then return false end
    end
    return true
  end)
  a.execute(function(unit, tile)
    edit.do_capture_units(unit, tile)
  end)
end)
```
Which offers a great deal more flexibility in how we set up rules and how we organise them in rulesets. That'd have to be down the road for now though.

**Comment by lmoureaux** [comment]
The AI needs to deduce things from the ruleset, so I don't think a Turing-complete language is a good idea. The proposal also makes ruleset a lot less editable by common mortals who don't speak programmerese.

Arbitrary logic expressions would already be a significant improvement. The syntax could match Lua:
```lua
reqs = unit.utype:has_flag("Capturable") and ... or...
```
But even this is already weird with all the `.` and `:`.

For effects, we could have the current tabulation-based approach or something formula-oriented:
```lua
value = city:size() * 2 + 5
```

**Comment by XHawk87** [comment]
> The AI needs to deduce things from the ruleset, so I don't think a Turing-complete language is a good idea.
Can you explain what you mean by this? The AI can run the `allow_when` function to know when it's allowed to use an action. What's missing? Some indication of when the outcome of the action would be desirable for the AI? 

> The proposal also makes ruleset a lot less editable by common mortals who don't speak programmerese.
> 
> Arbitrary logic expressions would already be a significant improvement. The syntax could match Lua:
> 
> reqs = unit.utype:has_flag("Capturable") and ... or...
> But even this is already weird with all the `.` and `:`.
> 
> For effects, we could have the current tabulation-based approach or something formula-oriented:
> 
> value = city:size() * 2 + 5

Well, I see no problem with having two methods of doing this: a simplified one for normal users and an advanced one using Lua for modders.

For the simplified one, to avoid using funny symbols we could limit it to only standard arithmetic and provide a set of simpified functions. E.g. `reqs = unit_type_has_flag("Capturable") and ... or ...`

**Comment by XHawk87** [comment]
> For instance, when deciding whether to research some tech the AI checks whether it can actually use the things provided by the tech. This way a land-locked AI doesn't research boats. It also does checks like this when deciding what to build:
> 
> > Check if all the requirements for the currently evaluated effect are met, except for having the building that we are evaluating.
> 
> It would be very hard to teach the AI to satisfy requirements if you can't reason about them.

It should be possible to pose a what-if to a Lua function, since we set the context for its evaluation. We'd just need a layer of alterations to the current game environment that the AI is considering which intercepts any requests for information that the Lua function makes, and falls back to the actual state of the game if there aren't any. 

If we're adding complex new functionality as part of a ruleset though, it's on the understanding that the AI won't know how to use it. If we're defining it in Lua though, we also have the opportunity to provide tools for ruleset modders to instruct the AI on how to use ruleset-specific features if so desired.

> The governor and the autogenerated help are other examples that would be harmed arbitrarily complexity in requirements.
For the governor, same answer as above. 

The auto-generated help text is something that could do with some work in any case, we already have a confusing mish-mash of handwritten and generated notes that don't always agree with eachother, and the current system of having a "quiet" column in tabular format to determine if a note should be added to a particular help file is I think poorly understood and frequently overlooked.

It might make sense to have a way to define help text in a declarative manner within the ruleset. E.g. 
```lua
action.create("capture_units", _("Capture Units"), function(a)
  ...
  a:add_help_text(_("This action allows the unit to capture unprotected enemy units. blah blah blah"))
  a:add_help_for_unit_flag("Capturer", _("Can capture units"))
  a:add_help_for_unit_flag("Capturable", _("Can be captured"))
  a:add_help_for_unit_flag("WeakCapturable", _("Can be captured when below 5 hitpoints"))
end)
```
I don't think we even have dedicated help for actions at the moment, but it seems like a good feature. 

While also allowing us to write things that are more readable than the auto-generated help would do. I don't know about you but I find this a bit hard to follow:

![Image](https://github.com/user-attachments/assets/29938ae1-ea21-454c-91c6-09a1a15c7e67)

This could be a much simpler list showing one-line rules for the unit, linking to other pages that give full explanations of how that particular Class, Effect, Action etc works.

**Comment by XHawk87** [comment]
> > It should be possible to pose a what-if to a Lua function, since we set the context for its evaluation. We'd just need a layer of alterations to the current game environment that the AI is considering which intercepts any requests for information that the Lua function makes, and falls back to the actual state of the game if there aren't any.
> 
> This makes queries like "how do I get a defense bonus" a whole lot more complicated, no?

Well, in the case of defence bonuses, there's already a lot of AI stupidity going on when doing that. In Sim06 for example, with the introduction of defence penalties for unit stack sizes, the AI kept piling on more units into its cities even though it was lowering its overall defence. If this approach had been used instead, of looking at future consequences for specific decisions, it would have been able to make a more sensible decision. 

To simplify AI decision-making however, we could provide scores for any given action, technology etc. E.g.
```lua
action.create("capture_units", _("Capture Units"), function(a)
  ...
  a:ai_can_use_score(function(player)
    ...
  end)
  a:ai_used_against_score(function(player)
    ...
  end)
end)
```
This has the advantage of placing all relevant code together in one location, all the rules about capturing units, when you can do it, what the AI thinks about it, the help text for it etc in one place, allowing us to have good separation of concerns with lower code fragmentation.

**Comment by lmoureaux** [comment]
Ok this might be a good idea. Let's open a new issue to discuss it and focus on the OR here... Ideally it comes sooner than full Lua flexibility.
